src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+  rwa [ker_one, quotient_group.ker_mk] at h,
+end
+
+end
+
+end monoid_hom
+
+section
+
+open category_theory
+
+namespace Group
+
+variables {A B : Group.{u}} (f : A ⟶ B)
+
+@[to_additive AddGroup.ker_eq_bot_of_mono]
+lemma ker_eq_bot_of_mono [mono f] : f.ker = ⊥ :=
+monoid_hom.ker_eq_bot_of_cancel $ λ u v,
+  (@cancel_mono _ _ _ _ _ f _
+    (show Group.of f.ker ⟶ A, from u) _).1
+
+@[to_additive AddGroup.mono_iff_ker_eq_bot]
+lemma mono_iff_ker_eq_bot :
+  mono f ↔ f.ker = ⊥ :=
+⟨λ h, @@ker_eq_bot_of_mono f h,
+λ h, concrete_category.mono_of_injective _ $ (monoid_hom.ker_eq_bot_iff f).1 h⟩
+
+@[to_additive AddGroup.mono_iff_injective]
+lemma mono_iff_injective :
+  mono f ↔ function.injective f :=
+iff.trans (mono_iff_ker_eq_bot f) $ monoid_hom.ker_eq_bot_iff f
+
+namespace surjective_of_epi_auxs
+
+local notation `X` := set.range (function.swap left_coset f.range.carrier)
+
+/--
+Define `X'` to be the set of all right cosets with an extra point at "infinity".
+-/
+@[nolint has_inhabited_instance]
+inductive X_with_infinity
+| from_coset : set.range (function.swap left_coset f.range.carrier) → X_with_infinity
+| infinity : X_with_infinity
+
+
+open X_with_infinity equiv.perm
+open_locale coset
+
+local notation `X'` := X_with_infinity f
+local notation `∞` := X_with_infinity.infinity
+local notation `SX'` := equiv.perm X'
+
+instance : has_smul B X' :=
+{ smul := λ b x, match x with
+  | from_coset y := from_coset ⟨b *l y, begin
+    rw [←subtype.val_eq_coe, ←y.2.some_spec, left_coset_assoc],
+    use b * y.2.some,
+  end⟩
+  | ∞ := ∞
+  end }
+
+lemma mul_smul (b b' : B) (x : X') :
+  (b * b') • x = b • b' • x :=
+match x with
+| from_coset y := begin
+  change from_coset _ = from_coset _,
+  simp only [←subtype.val_eq_coe, left_coset_assoc],
+end
+| ∞ := rfl
+end
+
+lemma one_smul (x : X') :
+  (1 : B) • x = x :=
+match x with
+| from_coset y := begin
+  change from_coset _ = from_coset _,
+  simp only [←subtype.val_eq_coe, one_left_coset, subtype.ext_iff_val],
+end
+| ∞ := rfl
+end
+
+lemma from_coset_eq_of_mem_range {b : B} (hb : b ∈ f.range) :
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ =
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+begin
+  congr,
+  change b *l f.range = f.range,
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm],
+  rw [left_coset_eq_iff, mul_one],
+  exact subgroup.inv_mem _ hb,
+end
+
+lemma from_coset_ne_of_nin_range {b : B} (hb : b ∉ f.range) :
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ ≠
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+begin
+  intros r,
+  simp only [subtype.mk_eq_mk] at r,
+  change b *l f.range = f.range at r,
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm] at r,
+  rw [left_coset_eq_iff, mul_one] at r,
+  exact hb (inv_inv b ▸ (subgroup.inv_mem _ r)),
+end
+
+
+noncomputable instance : decidable_eq X' := classical.dec_eq _
```

Comment:
You can just put
```lean
open_locale classical
noncomputable theory
```
at the top of the file and remove these decidability instances and all `noncomputable` declarations.